src/dune_rules/odoc_new.ml

```diff
@@ -1835,28 +1839,45 @@ let hierarchical_index_rules sctx ~all (tree : Index_tree.info Index_tree.t) =
 
 let hierarchical_html_rules sctx all tree =
   let ctx = Super_context.context sctx in
-  Index_tree.fold ~init:(Memo.return []) tree ~f:(fun index (ii : Index_tree.info) dirs ->
-    let* dirs = dirs in
-    let artifacts =
+  Index_tree.fold
+    ~init:(Memo.return (None, []))
+    tree
+    ~f:(fun index (ii : Index_tree.info) acc ->
+      let* search_db, dirs = acc in
       let index_artifact = Artifact.index ctx ~all:true index in
-      let all_artifacts = Index_tree.all_artifacts ii in
-      List.filter ~f:Artifact.is_visible (index_artifact :: all_artifacts)
-    in
-    let* new_dirs =
-      Memo.List.filter_map artifacts ~f:(fun a -> html_generate sctx true a)
-    in
-    let+ () =
-      let html_files =
-        List.filter artifacts ~f:Artifact.is_visible
-        |> List.map ~f:(fun a -> Path.build (Artifact.html_file a))
+      let artifacts =
+        let all_artifacts = Index_tree.all_artifacts ii in
+        List.filter ~f:Artifact.is_visible (index_artifact :: all_artifacts)
       in
-      let html_alias =
-        let html_dir = Index.html_dir ctx ~all index in
-        Dep.html_alias html_dir
+      let* search_db =
+        match ii.package with
+        | Some _pkg ->
+          let dir = Artifact.html_dir index_artifact in
+          let* sherlodoc_installed = Sherlodoc.is_installed sctx ~dir in
+          if sherlodoc_installed
+          then (
+            let odocls = List.map ~f:Artifact.odocl_file artifacts in
+            let+ () = Sherlodoc.search_db sctx ~dir odocls in
+            Some (Sherlodoc.Paths.db_dot_js ~dir))
+          else Memo.return None
+        | None -> Memo.return search_db
       in
-      Rules.Produce.Alias.add_deps html_alias (Action_builder.paths html_files)
-    in
-    new_dirs @ dirs)
+      let* new_dirs =
+        Memo.List.filter_map artifacts ~f:(fun a -> html_generate sctx ~search_db true a)
+      in
+      let+ () =
+        let html_files =
+          List.filter artifacts ~f:Artifact.is_visible
+          |> List.map ~f:(fun a -> Path.build (Artifact.html_file a))
```

Comment:
`filter_map`

Reply:
I agree filter_map is better, its just that I did not write this code, only the indentation has changed.